File: python/external-user-service/OrderService/services/order_service.py
"""
Order service implementation
Equivalent to the .NET OrderService.Services.OrderService
"""

import logging
from datetime import datetime
from decimal import Decimal
from typing import List, Optional
import uuid
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, and_
from sqlalchemy.orm import selectinload

from models.entities import Order, OrderItem, Customer, Product
from models.dtos import OrderDto, CreateOrderDto, UpdateOrderDto, CustomerDto, OrderItemDto, ProductDto

logger = logging.getLogger(__name__)


class OrderService:
    """Order service for managing order operations"""
    
    def __init__(self, db_session: AsyncSession):
        self.db_session = db_session
    
    async def get_all_orders(self) -> List[OrderDto]:
        """Get all orders with customer and order items"""
        result = await self.db_session.execute(
            select(Order)
            .options(
                selectinload(Order.customer),
                selectinload(Order.order_items).selectinload(OrderItem.product)
            )
            .order_by(Order.order_date.desc())
        )
        orders = result.scalars().all()
        return [self._map_to_dto(order) for order in orders]
    
    async def get_order_by_id(self, order_id: int) -> Optional[OrderDto]:
        """Get order by ID with customer and order items"""
        result = await self.db_session.execute(
            select(Order)
            .options(
                selectinload(Order.customer),
                selectinload(Order.order_items).selectinload(OrderItem.product)
            )
            .where(Order.id == order_id)
        )
        order = result.scalar_one_or_none()
        return self._map_to_dto(order) if order else None
    
    async def create_order(self, create_dto: CreateOrderDto) -> OrderDto:
        """Create a new order"""
        # Validate customer exists
        customer_result = await self.db_session.execute(
            select(Customer).where(Customer.id == create_dto.customer_id)
        )
        customer = customer_result.scalar_one_or_none()
        
        if not customer:
            raise ValueError("Customer not found")
        
        # Validate products and calculate totals
        sub_total = Decimal("0.00")
        order_items_data = []
        
        for item_dto in create_dto.order_items:
            # Get product
            product_result = await self.db_session.execute(
                select(Product).where(Product.id == item_dto.product_id)
            )
            product = product_result.scalar_one_or_none()
            
            if not product:
                raise ValueError(f"Product with ID {item_dto.product_id} not found")
            
            if not product.is_active:
                raise ValueError(f"Product {product.name} is not active")
            
            if product.stock_quantity < item_dto.quantity:
                raise ValueError(f"Insufficient stock for product {product.name}")
            
            # Calculate item total
            item_total = product.price * item_dto.quantity
            sub_total += item_total
            
            order_items_data.append({
                "product": product,
                "quantity": item_dto.quantity,
                "unit_price": product.price,
                "total_price": item_total
            })
        
        # Calculate tax and shipping (simplified calculation)
        tax_amount = sub_total * Decimal("0.08")  # 8% tax
        shipping_amount = Decimal("10.00") if sub_total < Decimal("100.00") else Decimal("0.00")
        total_amount = sub_total + tax_amount + shipping_amount
        
        # Generate order number
        order_number = f"ORD-{datetime.utcnow().strftime('%Y%m%d')}-{str(uuid.uuid4())[:8].upper()}"
        
        # Create order
        order = Order(
            customer_id=create_dto.customer_id,
            order_number=order_number,
            order_date=datetime.utcnow(),
            status="Pending",
            sub_total=sub_total,
            tax_amount=tax_amount,
            shipping_amount=shipping_amount,
            total_amount=total_amount,
            notes=create_dto.notes,
            shipping_address=create_dto.shipping_address,
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow()
        )
        
        self.db_session.add(order)
        await self.db_session.flush()  # Get the order ID
        
        # Create order items and update stock
        for item_data in order_items_data:
            order_item = OrderItem(
                order_id=order.id,
                product_id=item_data["product"].id,
                quantity=item_data["quantity"],
                unit_price=item_data["unit_price"],
                total_price=item_data["total_price"],
                created_at=datetime.utcnow(),
                updated_at=datetime.utcnow()
            )
            self.db_session.add(order_item)
            
            # Update product stock
            item_data["product"].stock_quantity -= item_data["quantity"]
            item_data["product"].updated_at = datetime.utcnow()
        
        await self.db_session.commit()
        
        # Refresh and return
        await self.db_session.refresh(order)
        result = await self.db_session.execute(
            select(Order)
            .options(
                selectinload(Order.customer),
                selectinload(Order.order_items).selectinload(OrderItem.product)
            )
            .where(Order.id == order.id)
        )
        order = result.scalar_one()
        
        logger.info(f"Order created with ID: {order.id}, Number: {order.order_number}")
        return self._map_to_dto(order)
    
    async def update_order(self, order_id: int, update_dto: UpdateOrderDto) -> Optional[OrderDto]:
        """Update an existing order"""
        # Get existing order
        result = await self.db_session.execute(
            select(Order)
            .options(
                selectinload(Order.customer),
                selectinload(Order.order_items).selectinload(OrderItem.product)
            )
            .where(Order.id == order_id)
        )
        order = result.scalar_one_or_none()
        
        if not order:
            return None
        
        # Update order
        order.status = update_dto.status
        order.notes = update_dto.notes
        order.shipping_address = update_dto.shipping_address
        order.updated_at = datetime.utcnow()
        
        await self.db_session.commit()
        await self.db_session.refresh(order)
        
        logger.info(f"Order updated with ID: {order_id}")
        return self._map_to_dto(order)
    
    async def update_order_status(self, order_id: int, status: str) -> Optional[OrderDto]:
        """Update order status"""
        valid_statuses = ["Pending", "Processing", "Shipped", "Delivered", "Cancelled"]
        if status not in valid_statuses:
            raise ValueError(f"Invalid status. Must be one of: {', '.join(valid_statuses)}")
        
        # Get existing order
        result = await self.db_session.execute(
            select(Order)
            .options(
                selectinload(Order.customer),
                selectinload(Order.order_items).selectinload(OrderItem.product)
            )
            .where(Order.id == order_id)
        )
        order = result.scalar_one_or_none()
        
        if not order:
            return None
        
        # Update status
        order.status = status
        order.updated_at = datetime.utcnow()
        
        await self.db_session.commit()
        await self.db_session.refresh(order)
        
        logger.info(f"Order status updated to '{status}' for order ID: {order_id}")
        return self._map_to_dto(order)
    
    async def delete_order(self, order_id: int) -> bool:
        """Delete an order"""
        # Get order
        result = await self.db_session.execute(
            select(Order)
            .options(selectinload(Order.order_items))
            .where(Order.id == order_id)
        )
        order = result.scalar_one_or_none()
        
        if not order:
            return False
        
        # Check if order can be deleted (only pending orders)
        if order.status not in ["Pending"]:
            raise RuntimeError("Cannot delete order that is not in Pending status")
        
        # Restore product stock
        for order_item in order.order_items:
            product_result = await self.db_session.execute(
                select(Product).where(Product.id == order_item.product_id)
            )
            product = product_result.scalar_one_or_none()
            if product:
                product.stock_quantity += order_item.quantity
                product.updated_at = datetime.utcnow()
        
        # Delete order (order items will be deleted due to cascade)
        await self.db_session.delete(order)
        await self.db_session.commit()
        
        logger.info(f"Order deleted with ID: {order_id}")
        return True
    
    async def get_orders_by_customer(self, customer_id: int) -> List[OrderDto]:
        """Get orders by customer ID"""
        result = await self.db_session.execute(
            select(Order)
            .options(
                selectinload(Order.customer),
                selectinload(Order.order_items).selectinload(OrderItem.product)
            )
            .where(Order.customer_id == customer_id)
            .order_by(Order.order_date.desc())
        )
        orders = result.scalars().all()
        return [self._map_to_dto(order) for order in orders]
    
    def _map_to_dto(self, order: Order) -> OrderDto:
        """Map Order entity to OrderDto"""
        return OrderDto(
            id=order.id,
            customer_id=order.customer_id,
            order_number=order.order_number,
            order_date=order.order_date,
            status=order.status,
            sub_total=order.sub_total,
            tax_amount=order.tax_amount,
            shipping_amount=order.shipping_amount,
            total_amount=order.total_amount,
            notes=order.notes,
            shipping_address=order.shipping_address,
            created_at=order.created_at,
            updated_at=order.updated_at,
            customer=CustomerDto(
                id=order.customer.id,
                first_name=order.customer.first_name,
                last_name=order.customer.last_name,
                email=order.customer.email,
                phone=order.customer.phone,
                address=order.customer.address,
                created_at=order.customer.created_at,
                updated_at=order.customer.updated_at
            ),
            order_items=[
                OrderItemDto(
                    id=item.id,
                    order_id=item.order_id,
                    product_id=item.product_id,
                    quantity=item.quantity,
                    unit_price=item.unit_price,
                    total_price=item.total_price,
                    created_at=item.created_at,
                    updated_at=item.updated_at,
                    product=ProductDto(
                        id=item.product.id,
                        name=item.product.name,
                        description=item.product.description,
                        price=item.product.price,
                        sku=item.product.sku,
                        stock_quantity=item.product.stock_quantity,
                        category=item.product.category,
                        is_active=item.product.is_active,
                        created_at=item.product.created_at,
                        updated_at=item.product.updated_at
                    )
                )
                for item in order.order_items
            ]
        )

File: nodejs/external-user-service/OrderService/src/services/order.service.ts
/**
 * Order service for Order Service
 * TypeScript equivalent to Python order service
 */

import { Order, OrderItem, Customer, Product } from '../models/entities';
import { CreateOrderDto, UpdateOrderDto, OrderDto, VALID_ORDER_STATUSES, OrderStatus } from '../models/dtos';
import { ProductService } from './product.service';
import { logger } from '../common/logger';
import { v4 as uuidv4 } from 'uuid';

export class OrderService {
    private productService: ProductService;

    constructor() {
        this.productService = new ProductService();
    }

    async getAllOrders(): Promise<OrderDto[]> {
        try {
            const orders = await Order.findAll({
                include: [
                    {
                        model: Customer,
                        as: 'customer'
                    },
                    {
                        model: OrderItem,
                        as: 'order_items',
                        include: [{
                            model: Product,
                            as: 'product'
                        }]
                    }
                ],
                order: [['created_at', 'DESC']]
            });
            return orders.map(order => this.mapToDto(order));
        } catch (error) {
            logger.error('Error getting all orders:', error);
            throw error;
        }
    }

    async getOrderById(id: number): Promise<OrderDto> {
        try {
            const order = await Order.findByPk(id, {
                include: [
                    {
                        model: Customer,
                        as: 'customer'
                    },
                    {
                        model: OrderItem,
                        as: 'order_items',
                        include: [{
                            model: Product,
                            as: 'product'
                        }]
                    }
                ]
            });
            
            if (!order) {
                const error = new Error(`Order with ID ${id} not found`);
                (error as any).statusCode = 404;
                throw error;
            }
            
            return this.mapToDto(order);
        } catch (error) {
            logger.error(`Error getting order by ID ${id}:`, error);
            throw error;
        }
    }

    async createOrder(orderData: CreateOrderDto): Promise<OrderDto> {
        try {
            // Verify customer exists
            const customer = await Customer.findByPk(orderData.customer_id);
            if (!customer) {
                const error = new Error(`Customer with ID ${orderData.customer_id} not found`);
                (error as any).statusCode = 404;
                throw error;
            }

            // Calculate totals and validate products
            let subTotal = 0;
            const orderItems = [];

            for (const item of orderData.order_items) {
                const product = await Product.findByPk(item.product_id);
                if (!product || !product.is_active) {
                    const error = new Error(`Product with ID ${item.product_id} not found or inactive`);
                    (error as any).statusCode = 404;
                    throw error;
                }

                if (product.stock_quantity < item.quantity) {
                    const error = new Error(`Insufficient stock for product ${product.name}. Available: ${product.stock_quantity}, Required: ${item.quantity}`);
                    (error as any).statusCode = 400;
                    throw error;
                }

                const unitPrice = product.price;
                const totalPrice = unitPrice * item.quantity;
                subTotal += totalPrice;

                orderItems.push({
                    product_id: item.product_id,
                    quantity: item.quantity,
                    unit_price: unitPrice,
                    total_price: totalPrice
                });
            }

            // Calculate tax and shipping (simplified)
            const taxAmount = subTotal * 0.08; // 8% tax
            const shippingAmount = subTotal > 100 ? 0 : 10; // Free shipping over $100
            const totalAmount = subTotal + taxAmount + shippingAmount;

            // Generate order number
            const orderNumber = `ORD-${Date.now()}-${Math.random().toString(36).substr(2, 9).toUpperCase()}`;

            // Create order
            const order = await Order.create({
                customer_id: orderData.customer_id,
                order_number: orderNumber,
                status: 'Pending',
                sub_total: subTotal,
                tax_amount: taxAmount,
                shipping_amount: shippingAmount,
                total_amount: totalAmount,
                notes: orderData.notes,
                shipping_address: orderData.shipping_address
            });

            // Create order items and update product stock
            for (const itemData of orderItems) {
                await OrderItem.create({
                    order_id: order.id,
                    ...itemData
                });

                // Update product stock
                await this.productService.updateStock(itemData.product_id, itemData.quantity);
            }

            logger.info(`Order created with ID: ${order.id}`);
            return await this.getOrderById(order.id);
        } catch (error) {
            logger.error('Error creating order:', error);
            throw error;
        }
    }

    async updateOrder(id: number, orderData: UpdateOrderDto): Promise<OrderDto> {
        try {
            const order = await Order.findByPk(id);
            if (!order) {
                const error = new Error(`Order with ID ${id} not found`);
                (error as any).statusCode = 404;
                throw error;
            }

            // Validate status
            if (!VALID_ORDER_STATUSES.includes(orderData.status as OrderStatus)) {
                const error = new Error(`Invalid status. Valid statuses: ${VALID_ORDER_STATUSES.join(', ')}`);
                (error as any).statusCode = 400;
                throw error;
            }

            await order.update({
                status: orderData.status,
                notes: orderData.notes,
                shipping_address: orderData.shipping_address
            });

            logger.info(`Order updated with ID: ${id}`);
            return await this.getOrderById(id);
        } catch (error) {
            logger.error(`Error updating order with ID ${id}:`, error);
            throw error;
        }
    }

    async updateOrderStatus(id: number, status: string): Promise<OrderDto> {
        try {
            const order = await Order.findByPk(id);
            if (!order) {
                const error = new Error(`Order with ID ${id} not found`);
                (error as any).statusCode = 404;
                throw error;
            }

            // Validate status
            if (!VALID_ORDER_STATUSES.includes(status as OrderStatus)) {
                const error = new Error(`Invalid status. Valid statuses: ${VALID_ORDER_STATUSES.join(', ')}`);
                (error as any).statusCode = 400;
                throw error;
            }

            await order.update({ status });
            logger.info(`Order status updated for ID: ${id} to ${status}`);
            return await this.getOrderById(id);
        } catch (error) {
            logger.error(`Error updating order status for ID ${id}:`, error);
            throw error;
        }
    }

    async deleteOrder(id: number): Promise<void> {
        try {
            const order = await Order.findByPk(id, {
                include: [{
                    model: OrderItem,
                    as: 'order_items'
                }]
            });
            
            if (!order) {
                const error = new Error(`Order with ID ${id} not found`);
                (error as any).statusCode = 404;
                throw error;
            }

            // Only allow deletion of pending orders
            if (order.status !== 'Pending') {
                const error = new Error(`Cannot delete order with status: ${order.status}`);
                (error as any).statusCode = 400;
                throw error;
            }

            // Restore product stock
            for (const item of order.order_items) {
                const product = await Product.findByPk(item.product_id);
                if (product) {
                    await product.update({
                        stock_quantity: product.stock_quantity + item.quantity
                    });
                }
            }

            await order.destroy();
            logger.info(`Order deleted with ID: ${id}`);
        } catch (error) {
            logger.error(`Error deleting order with ID ${id}:`, error);
            throw error;
        }
    }

    private mapToDto(order: Order): OrderDto {
        return {
            id: order.id,
            customer_id: order.customer_id,
            order_number: order.order_number,
            order_date: order.order_date,
            status: order.status,
            sub_total: order.sub_total,
            tax_amount: order.tax_amount,
            shipping_amount: order.shipping_amount,
            total_amount: order.total_amount,
            notes: order.notes,
            shipping_address: order.shipping_address,
            created_at: order.created_at,
            updated_at: order.updated_at,
            customer: {
                id: order.customer.id,
                first_name: order.customer.first_name,
                last_name: order.customer.last_name,
                email: order.customer.email,
                phone: order.customer.phone,
                address: order.customer.address,
                created_at: order.customer.created_at,
                updated_at: order.customer.updated_at
            },
            order_items: order.order_items.map(item => ({
                id: item.id,
                order_id: item.order_id,
                product_id: item.product_id,
                quantity: item.quantity,
                unit_price: item.unit_price,
                total_price: item.total_price,
                created_at: item.created_at,
                updated_at: item.updated_at,
                product: {
                    id: item.product.id,
                    name: item.product.name,
                    description: item.product.description,
                    price: item.product.price,
                    sku: item.product.sku,
                    stock_quantity: item.product.stock_quantity,
                    category: item.product.category,
                    is_active: item.product.is_active,
                    created_at: item.product.created_at,
                    updated_at: item.product.updated_at
                }
            }))
        };
    }
}
